internal/lockclient/simple_client.go

```diff
@@ -103,3 +137,331 @@ func (sc *SimpleClient) Release(d lockservice.Descriptors) error {
+				curOwner, err := sc.Checkacquire(d)
+				if curOwner == "" && err != nil {
+					desc := lockservice.NewLockDescriptor(d.ObjectID, op)
+					err = sc.acquire(desc)
+					if err == nil {
+						sc.mu.Lock()
+						// Once the task of acquiring is complete, remove
+						// the first element from the slice as it was granted the lock.
+						sc.pouncersMap[d] = append(sc.pouncersMap[d][:0], sc.pouncersMap[d][1:]...)
+						sc.mu.Unlock()
+
+						log.Debug().
+							Str("lock granted through pounce to", op).
+							Msg("pounce success")
+
+						// Exit condition is the pouncer getting access to the lock.
+						if op == owner || len(sc.pouncers(d)) == 0 {
+							log.Debug().
+								Str("owner", owner).
+								Msg("stopping pouncing process, pouncer availed lock")
+							q <- struct{}{}
+							close(q)
+							return nil
+						}
+					}
+				}
+
+			}
+		}
+	}
+}
+
+// Pouncers returns the active "pouncers" on a descriptor.
+// This is safe to read concurrently.
+func (sc *SimpleClient) pouncers(d lockservice.ObjectDescriptor) []string {
+	sc.mu.Lock()
+	var x = make([]string, len(sc.pouncersMap[d]))
+	copy(x, sc.pouncersMap[d])
+	sc.mu.Unlock()
+	// fmt.Println(x)
+	// fmt.Println(sc.pouncers[d])
+	return x
+}
+
+// Checkacquire checks for acquisition of lock and returns the owner if the lock
+// is already acquired.
+// The errors returned can be due to HTTP errors or marshalling errors.
+// A "file is not acquired" error is returned if so and no error and an owner is
+// returned if the object is acquired.
+func (sc *SimpleClient) Checkacquire(d lockservice.ObjectDescriptor) (string, error) {
+	if sc.cache != nil {
+		owner, err := sc.getFromCache(d)
+		if err != nil {
+			return "", err
+		}
+		return owner, nil
+	}
+
+	endPoint := sc.config.IPAddr + ":" + sc.config.PortAddr + "/checkacquire"
+	data := lockservice.LockCheckRequest{FileID: d.ObjectID}
+	requestJSON, err := json.Marshal(data)
+	if err != nil {
+		return "", err
+	}
+
+	req, err := http.NewRequest("POST", endPoint, bytes.NewBuffer(requestJSON))
+	if err != nil {
+		return "", err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", err
+	}
+	if resp.StatusCode != 200 {
+		return "", lockservice.Error(strings.TrimSpace(string(body)))
+	}
+
+	var ownerData lockservice.CheckacquireRes
+	err = json.Unmarshal(body, &ownerData)
+	if err != nil {
+		return "", err
+	}
+
+	return ownerData.Owner, nil
+}
+
+// getFromCache checks the lock status on the descriptor in the cache.
+// This function returns an error if the cache doesn't exist or the
+// file is NOT acquired.
+func (sc *SimpleClient) getFromCache(d lockservice.ObjectDescriptor) (string, error) {
+	if sc.cache != nil {
+		owner, err := sc.cache.GetElement(cache.NewSimpleKey(d.ObjectID, ""))
```

Comment:
If `GetElement()` is returning the owner, we should change it such that passing `""` to it as a parameter for the owner of the element isn't needed. What do you think?

Reply:
Agreed, I think this was a quick fix by me. Maybe I should create a new object, Ill see what I can do.

Reply:
Actually, these are my options:
1. Create another function that does it (overloading is not allowed so yeah new function with a different sign is needed)
2. Pass `&SimpleKey{d.ObjectID}`

So the current way seemed to be the most clean one.